File: examples/demos/scenes/hybrid_dynamics.cpp
/*
 * Copyright (c) 2011, The DART development contributors
 * All rights reserved.
 *
 * The list of contributors can be found at:
 *   https://github.com/dartsim/dart/blob/main/LICENSE
 *
 * This file is provided under the "BSD-style" License.
 */

#include "scenes.hpp"
#include "z_up.hpp"

#include <dart/gui/panel.hpp>
#include <dart/gui/viewer.hpp>

#include <dart/simulation/world.hpp>

#include <dart/dynamics/body_node.hpp>
#include <dart/dynamics/joint.hpp>
#include <dart/dynamics/skeleton.hpp>

#include <dart/io/read.hpp>

#include <Eigen/Core>

#include <memory>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

#include <cmath>

namespace dart::examples::demos {

namespace {

constexpr const char* kHybridWorldUri = "dart://sample/skel/fullbody1.skel";
constexpr const char* kBipedSkeletonName = "fullbody1";

dart::dynamics::Joint* getRequiredJoint(
    const dart::dynamics::SkeletonPtr& skeleton, const char* name)
{
  auto* joint = skeleton == nullptr ? nullptr : skeleton->getJoint(name);
  if (joint == nullptr) {
    throw std::runtime_error(
        "hybrid_dynamics world is missing joint: " + std::string(name));
  }
  return joint;
}

std::size_t getRequiredCommandIndex(
    const dart::dynamics::SkeletonPtr& skeleton, const char* jointName)
{
  return getRequiredJoint(skeleton, jointName)->getIndexInSkeleton(0);
}

dart::simulation::WorldPtr createHybridDynamicsWorld()
{
  auto world = dart::io::readWorld(kHybridWorldUri);
  if (world == nullptr) {
    throw std::runtime_error(
        "Failed to load hybrid_dynamics world from "
        + std::string(kHybridWorldUri));
  }
  auto ground = world->getSkeleton("ground skeleton");
  if (ground == nullptr) {
    throw std::runtime_error("hybrid_dynamics world is missing ground");
  }

  auto biped = world->getSkeleton(kBipedSkeletonName);
  if (biped == nullptr) {
    throw std::runtime_error("hybrid_dynamics world is missing fullbody1");
  }

  const std::vector<std::size_t> genCoordIds{
      1,  // global orientation y
      6,  // left hip
      9,  // left knee
      10, // left ankle
      13, // right hip
      16, // right knee
      17, // right ankle
      21, // lower back
  };
  Eigen::VectorXd initConfig(8);
  initConfig << -0.2, 0.15, -0.4, 0.25, 0.15, -0.4, 0.25, 0.0;
  biped->setPositions(genCoordIds, initConfig);

  if (auto* rootJoint = biped->getJoint(0)) {
    rootJoint->setActuatorType(dart::dynamics::Joint::ActuatorType::PASSIVE);
  }
  for (std::size_t i = 1; i < biped->getNumJoints(); ++i) {
    if (auto* joint = biped->getJoint(i)) {
      joint->setActuatorType(dart::dynamics::Joint::ActuatorType::VELOCITY);
    }
  }

  // The biped world is authored Y-up; reorient to the canonical Z-up
  // convention. The controller only issues joint-space velocity commands, so
  // the rotated motion is identical. The camera is flipped to Z-up to match.
  reorientWorldToZUp(world);

  return world;
}

class HybridDynamicsController
{
public:
  explicit HybridDynamicsController(dart::simulation::WorldPtr world)
    : mWorld(std::move(world))
  {
    if (mWorld == nullptr) {
      throw std::runtime_error("hybrid_dynamics controller is missing world");
    }
    mBiped = mWorld->getSkeleton(kBipedSkeletonName);
    if (mBiped == nullptr) {
      throw std::runtime_error("hybrid_dynamics controller is missing biped");
    }

    mScapulaLeft = getRequiredCommandIndex(mBiped, "j_scapula_left");
    mScapulaRight = getRequiredCommandIndex(mBiped, "j_scapula_right");
    mForearmLeft = getRequiredCommandIndex(mBiped, "j_forearm_left");
    mForearmRight = getRequiredCommandIndex(mBiped, "j_forearm_right");
    mShinLeft = getRequiredCommandIndex(mBiped, "j_shin_left");
    mShinRight = getRequiredCommandIndex(mBiped, "j_shin_right");
  }

  void preStep()
  {
    const double armCycle = std::sin(mWorld->getTime() * 4.0);
    const double legCycle = std::sin(mWorld->getTime() * 2.0);
    mBiped->setCommand(mScapulaLeft, armCycle);
    mBiped->setCommand(mScapulaRight, -armCycle);
    mBiped->setCommand(mForearmLeft, 0.8 * armCycle);
    mBiped->setCommand(mForearmRight, 0.8 * armCycle);
    mBiped->setCommand(mShinLeft, 0.1 * legCycle);
    mBiped->setCommand(mShinRight, 0.1 * legCycle);
  }

  void toggleHarness()
  {
    auto* pelvis = mBiped->getBodyNode("h_pelvis");
    auto* joint = pelvis == nullptr ? nullptr : pelvis->getParentJoint();
    if (joint == nullptr) {
      return;
    }

    mHarnessOn = !mHarnessOn;
    joint->setActuatorType(
        mHarnessOn ? dart::dynamics::Joint::ActuatorType::LOCKED
                   : dart::dynamics::Joint::ActuatorType::PASSIVE);
  }

  bool harnessOn() const
  {
    return mHarnessOn;
  }

private:
  dart::simulation::WorldPtr mWorld;
  dart::dynamics::SkeletonPtr mBiped;
  std::size_t mScapulaLeft = 0;
  std::size_t mScapulaRight = 0;
  std::size_t mForearmLeft = 0;
  std::size_t mForearmRight = 0;
  std::size_t mShinLeft = 0;
  std::size_t mShinRight = 0;
  bool mHarnessOn = false;
};

dart::gui::OrbitCamera makeHybridDynamicsCamera()
{
  dart::gui::OrbitCamera camera;
  camera.target = Eigen::Vector3d::Zero();
  camera.up = Eigen::Vector3d::UnitZ();
  camera.yaw = 0.5404195002705842;
  camera.pitch = 0.4758822496604165;
  camera.distance = 6.557438524302;
  return camera;
}

std::vector<dart::gui::KeyboardAction> createHybridDynamicsKeyboardActions(
    const std::shared_ptr<HybridDynamicsController>& controller)
{
  dart::gui::KeyboardAction toggleHarness;
  toggleHarness.label = "Toggle hybrid-dynamics harness";
  toggleHarness.shortcut = dart::gui::KeyboardShortcut::characterKey('h');
  toggleHarness.callback = [controller](dart::gui::KeyboardActionContext&) {
    controller->toggleHarness();
  };
  return {std::move(toggleHarness)};
}

dart::gui::Panel createHybridDynamicsPanel(
    const std::shared_ptr<HybridDynamicsController>& controller)
{
  dart::gui::Panel panel;
  panel.title = "Hybrid Dynamics";
  panel.buildWithContext = [controller](
                               dart::gui::PanelBuilder& builder,
                               dart::gui::PanelContext& context) {
    builder.text("Scripted velocity commands on the fullbody biped");
    builder.text("'h': toggle harness on/off");
    builder.text("space bar: simulation on/off");
    builder.separator();
    if (context.lifecycle != nullptr) {
      if (builder.button(context.lifecycle->paused ? "Resume" : "Pause")) {
        dart::gui::togglePaused(*context.lifecycle);
      }
      builder.sameLine();
      if (builder.button("Step")) {
        dart::gui::requestSingleStep(*context.lifecycle);
      }
    }
    if (builder.button(
            controller->harnessOn() ? "Unlock pelvis" : "Lock pelvis")) {
      controller->toggleHarness();
    }
    builder.text(
        "harness: " + std::string(controller->harnessOn() ? "on" : "off"));
    builder.text("time: " + std::to_string(context.simulationTime));
    builder.text("contacts: " + std::to_string(context.contactCount));
  };
  return panel;
}

} // namespace

dart::gui::ApplicationOptions makeHybridDynamicsScene()
{
  auto world = createHybridDynamicsWorld();
  auto controller = std::make_shared<HybridDynamicsController>(world);

  dart::gui::ApplicationOptions options;
  options.world = world;
  options.camera = makeHybridDynamicsCamera();
  options.preStep = [controller]() {
    controller->preStep();
  };
  options.keyboardActions = createHybridDynamicsKeyboardActions(controller);
  options.panels.push_back(createHybridDynamicsPanel(controller));
  return options;
}

} // namespace dart::examples::demos

File: python/examples/demos/scenes/hybrid_dynamics.py
"""Hybrid dynamics: VELOCITY-actuated fullbody biped with PASSIVE root.

Mirrors examples/demos/scenes/hybrid_dynamics.cpp. Loads the
fullbody1 skeleton from dart://sample/skel/fullbody1.skel, sets the root
joint to PASSIVE and the rest to VELOCITY actuation, then drives
shoulder/forearm/shin joints with sinusoidal velocity commands.
"""

from __future__ import annotations

from collections import deque
import math

import numpy as np

import dartpy as dart

from ..runner import PythonDemoScene, ScenePanel, SceneSetup
from ._z_up import reorient_to_z_up

_WORLD_URI = "dart://sample/skel/fullbody1.skel"
_BIPED_NAME = "fullbody1"
_DRIVEN_JOINTS = (
    "j_scapula_left",
    "j_scapula_right",
    "j_forearm_left",
    "j_forearm_right",
    "j_shin_left",
    "j_shin_right",
)


def _required_joint(skel: "dart.Skeleton", joint_name: str) -> "dart.Joint":
    joint = skel.get_joint(joint_name)
    if joint is None:
        raise RuntimeError(f"hybrid_dynamics world is missing joint: {joint_name}")
    return joint


def build() -> SceneSetup:
    world = dart.io.SkelParser.read_world(_WORLD_URI)
    if world is None:
        raise RuntimeError(f"Failed to load hybrid_dynamics world from {_WORLD_URI}")
    reorient_to_z_up(world)

    biped = world.get_skeleton(_BIPED_NAME)
    if biped is None:
        raise RuntimeError("hybrid_dynamics world is missing fullbody1")

    positions = np.array(biped.get_positions(), dtype=float)
    gen_coord_ids = [1, 6, 9, 10, 13, 16, 17, 21]
    init_config = [-0.2, 0.15, -0.4, 0.25, 0.15, -0.4, 0.25, 0.0]
    for index, value in zip(gen_coord_ids, init_config):
        positions[index] = value
    biped.set_positions(positions)

    root = biped.get_joint(0)
    if root is not None:
        root.set_actuator_type(dart.PASSIVE)
    for i in range(1, biped.get_num_joints()):
        joint = biped.get_joint(i)
        if joint is not None:
            joint.set_actuator_type(dart.VELOCITY)

    driven = {name: _required_joint(biped, name) for name in _DRIVEN_JOINTS}
    state = {"arm_cycle": 0.0, "leg_cycle": 0.0}
    arm_history = deque(maxlen=120)
    leg_history = deque(maxlen=120)
    com_z_history = deque(maxlen=120)

    def pre_step() -> None:
        t = world.get_time()
        arm_cycle = math.sin(t * 4.0)
        leg_cycle = math.sin(t * 2.0)
        state["arm_cycle"] = arm_cycle
        state["leg_cycle"] = leg_cycle
        driven["j_scapula_left"].set_command(0, arm_cycle)
        driven["j_scapula_right"].set_command(0, -arm_cycle)
        driven["j_forearm_left"].set_command(0, 0.8 * arm_cycle)
        driven["j_forearm_right"].set_command(0, 0.8 * arm_cycle)
        driven["j_shin_left"].set_command(0, 0.1 * leg_cycle)
        driven["j_shin_right"].set_command(0, 0.1 * leg_cycle)

    def build_panel(builder: object, context: object) -> None:
        com = np.asarray(biped.get_com(), dtype=float)
        arm_history.append(float(state["arm_cycle"]))
        leg_history.append(float(state["leg_cycle"]))
        com_z_history.append(float(com[2]))
        builder.text("actuation: passive root + velocity joints")
        builder.text(f"arm velocity command: {state['arm_cycle']:.3f} rad/s")
        builder.text(f"leg velocity command: {0.1 * state['leg_cycle']:.3f} rad/s")
        builder.text(f"COM height: {com[2]:.3f} m")
        builder.separator()
        builder.plot_lines("Arm command", list(arm_history))
        builder.plot_lines("Leg command", list(leg_history))
        builder.plot_lines("COM height", list(com_z_history))

    return SceneSetup(
        world=world,
        pre_step=pre_step,
        panels=[ScenePanel("Hybrid Dynamics", build_panel)],
        info={"driven_joints": list(_DRIVEN_JOINTS)},
    )


SCENE = PythonDemoScene(
    id="hybrid_dynamics",
    title="Hybrid Dynamics",
    category="Control & IK",
    summary="Scripted velocity commands on a fullbody biped with PASSIVE root.",
    build=build,
)
